File: clients/go/pkg/scheduling.go
package internal

import (
	"time"
)

func IsScheduledFeatureActive(feature Feature) bool {
	return IsScheduleActive(feature.Schedule, feature.ScheduleType)
}

func IsScheduleActive(schedule Schedule, scheduleType ScheduleType) bool {
	return IsScheduleActiveWithNow(schedule, scheduleType, time.Now().UTC())
}

func IsScheduleActiveWithNow(schedule Schedule, scheduleType ScheduleType, now time.Time) bool {

	switch scheduleType {
	case EmptyScheduleType:
		return true
	case GlobalScheduleType, EnvironmentScheduleType:

		startDate := time.UnixMilli(schedule.Start)
		startOfStartDate := time.Date(
			startDate.Year(),
			startDate.Month(),
			startDate.Day(),
			0,
			0,
			0,
			0,
			time.UTC,
		)
		endDate := time.UnixMilli(schedule.End)
		endOfEndDate := time.Date(
			endDate.Year(),
			endDate.Month(),
			endDate.Day(),
			23,
			59,
			59,
			999_999_999,
			time.UTC,
		)

		if !now.After(startOfStartDate) || !now.Before(endOfEndDate) {
			return false
		}

		startTime := time.UnixMilli(schedule.StartTime).UTC()
		endTime := time.UnixMilli(schedule.EndTime).UTC()

		switch schedule.TimeType {
		case NoneScheduleTimeType:
			return true
		case StartEndScheduleTimeType:
			isAfterStartDateTime := now.After(time.Date(
				startDate.Year(),
				startDate.Month(),
				startDate.Day(),
				startTime.Hour(),
				startTime.Minute(),
				startTime.Second(),
				startTime.Nanosecond(),
				time.UTC,
			))

			isBeforeEndDateTime := now.Before(time.Date(
				endDate.Year(),
				endDate.Month(),
				endDate.Day(),
				endTime.Hour(),
				endTime.Minute(),
				endTime.Second(),
				endTime.Nanosecond(),
				time.UTC,
			))
			return isAfterStartDateTime && isBeforeEndDateTime
		case DailyScheduleTimeType:

			zeroDay := time.UnixMilli(0).UTC()
			nowTimestamp := now.UnixMilli()

			todayZeroTimestamp := time.Date(
				now.Year(),
				now.Month(),
				now.Day(),
				0,
				0,
				0,
				0,
				time.UTC,
			).UnixMilli()

			zeroedStartTimestamp := time.Date(
				zeroDay.Year(),
				zeroDay.Month(),
				zeroDay.Day(),
				startTime.Hour(),
				startTime.Minute(),
				startTime.Second(),
				startTime.Nanosecond(),
				time.UTC,
			).UnixMilli()

			zeroedEndDateTime := time.Date(
				zeroDay.Year(),
				zeroDay.Month(),
				zeroDay.Day(),
				endTime.Hour(),
				endTime.Minute(),
				endTime.Second(),
				endTime.Nanosecond(),
				time.UTC,
			)

			zeroedEndTimestamp := zeroedEndDateTime.UnixMilli()

			startTimestamp := todayZeroTimestamp + zeroedStartTimestamp
			endTimestamp := todayZeroTimestamp + zeroedEndTimestamp

			if zeroedStartTimestamp > zeroedEndTimestamp {
				return nowTimestamp >= startTimestamp || nowTimestamp <= endTimestamp
			} else {
				return nowTimestamp >= startTimestamp && nowTimestamp <= endTimestamp
			}
		default:
			return false
		}
	default:
		return false
	}
}

File: clients/js/src/scheduling.ts
import dayjs from "dayjs";
import utc from "dayjs/plugin/utc";
dayjs.extend(utc);

import {
  VexillaFeature,
  VexillaSchedule,
  VexillaScheduleType,
} from "@vexilla/types";

export function isScheduledFeatureActive(feature: VexillaFeature) {
  return isScheduleActive(feature.schedule, feature.scheduleType);
}

export function isScheduleActive(
  schedule: VexillaSchedule,
  scheduleType: VexillaScheduleType
) {
  const now = Date.now();
  return isScheduleActiveWithNow(schedule, scheduleType, now);
}

export function isScheduleActiveWithNow(
  schedule: VexillaSchedule,
  scheduleType: VexillaScheduleType,
  now: number
) {
  if (!scheduleType) {
    return true;
  }

  const currentTime = dayjs.utc(now);

  const startOfStartDate = dayjs.utc(schedule.start).startOf("day");
  const endOfEndDate = dayjs.utc(schedule.end).endOf("day");

  if (
    currentTime.isBefore(startOfStartDate) ||
    currentTime.isAfter(endOfEndDate)
  ) {
    return false;
  }

  if (schedule.timeType === "none") {
    return true;
  }

  const startTime = dayjs.utc(schedule.startTime);
  const endTime = dayjs.utc(schedule.endTime);

  if (schedule.timeType === "start/end") {
    const startOfEndDate = dayjs.utc(schedule.end).startOf("day");

    const startDateTimestampWithStartTime =
      startOfStartDate.unix() * 1000 + schedule.startTime;
    const endDateTimestampWithEndTime =
      startOfEndDate.unix() * 1000 + schedule.endTime;

    return (
      startDateTimestampWithStartTime <= now &&
      now <= endDateTimestampWithEndTime
    );
  } else if (schedule.timeType === "daily") {
    const zeroDay = dayjs.utc(0);
    let zeroDayPlusCurrentTime = zeroDay
      .set("hour", currentTime.hour())
      .set("minute", currentTime.minute())
      .set("second", currentTime.second())
      .set("millisecond", currentTime.millisecond());

    let zeroedStart = zeroDay
      .set("hour", startTime.hour())
      .set("minute", startTime.minute())
      .set("second", startTime.second())
      .set("millisecond", startTime.millisecond());

    let zeroedEnd = zeroDay
      .set("hour", endTime.hour())
      .set("minute", endTime.minute())
      .set("second", endTime.second())
      .set("millisecond", endTime.millisecond());

    if (zeroedEnd.isBefore(zeroedStart)) {
      return (
        zeroDayPlusCurrentTime.isAfter(zeroedStart) ||
        zeroDayPlusCurrentTime.isBefore(zeroedEnd)
      );
    } else {
      return (
        zeroDayPlusCurrentTime.isAfter(zeroedStart) &&
        zeroDayPlusCurrentTime.isBefore(zeroedEnd)
      );
    }
  }

  return false;
}
